Update compression metrics

mogres opened the pull request:
Time Estimate or Size
---------------
Medium/Large

Updates the compression metric analysis to match other analyses in the repository.


Change summary:
---------------
* Refactor compression metric analysis
* Calculate metrics on individual repeats instead of a combined dataframe
* Add and update documentation


Steps to verify:
-----------------------
1. Run `subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py`
The sampled readdy dataframes are currently unavailable, but the workflow runs as expected on cytosim data.

jessicasyu committed:
Remove unused velocities and add random seed to template

jessicasyu committed:
Update run cytosim simulation notebooks

jessicasyu committed:
Update batch simulation workflows to save logs

jessicasyu committed:
Add docstrings to run batch simulations

jessicasyu committed:
Add io-collection to dependencies

jessicasyu committed:
Update methods for saving files to bucket

jessicasyu committed:
Move templates and configs to Cytosim repo

jessicasyu committed:
Add python-dotenv to dependencies

jessicasyu committed:
Update simulation workflow paths to use env

jessicasyu committed:
Rename run batch simulations file

mogres committed:
refactoring analysis part 2

mogres committed:
wrap metric calculation into a single function

mogres committed:
Merge branch 'main' of github.com:simularium/subcell-analysis into update-analysis

mogres committed:
Merge branch 'main' of github.com:simularium/subcell-analysis into update-compression-metrics

mogres committed:
Revert dependencies to main

mogres committed:
Merge branch 'main' of github.com:simularium/subcell-analysis into update-compression-metrics

mogres committed:
calculate metrics on individual repeats; remove unused functions from compression_analysis

mogres committed:
add readme for compression analysis

mogres committed:
Separate code and markdown blocks; add step to combine simulators; add parameter to recalculate metrics

mogres committed:
update docstrings; move vector functions, constants, and plotting tools to separate files

mogres committed:
Update autodoc generator scripts

mogres committed:
Revert requirements to main

mogres committed:
Remove unused init file

jessicasyu commented on docs/analysis.rst:
```diff
@@ -0,0 +1,14 @@
+Analysis workflow notebooks
+=============================
```
nit:

```suggestion
===========================
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/README.md:
```diff
@@ -0,0 +1,7 @@
+# Compression Metric Calculations
```
```suggestion
# Compression metric calculations
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/README.md:
```diff
@@ -0,0 +1,7 @@
+# Compression Metric Calculations
+
+## Metrics for Comparing Traces of Compressed Fibers
```
```suggestion
## Metrics for comparing traces of compressed fibers
```

Sentence case is easier to read than title case

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py:
```diff
@@ -0,0 +1,132 @@
+- [Load dataframes](#load-dataframes)
+- [Calculate metrics and add to dataframe](#calculate-metrics)
+- [Plot metrics vs time](#plot-metrics-vs-time)
+"""
```
Fix TOC links

mogres replied:
Fixed in [384b16](https://github.com/simularium/subcell-pipeline/pull/58/commits/384b16b4ca7645013baeb7b217276296ce9d9128)

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py:
```diff
@@ -0,0 +1,132 @@
+
+# %% [markdown]
+"""
+## Calculate metrics for cytosim data
```
```suggestion
## Calculate metrics for Cytosim data
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py:
```diff
@@ -0,0 +1,132 @@
+
+# %% [markdown]
+"""
+## Calculate metrics for readdy data
```
```suggestion
## Calculate metrics for ReaDDy data
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py:
```diff
@@ -0,0 +1,132 @@
+condition_keys: list[str] = ["0047", "0150", "0470", "1500"]
+
+# Location to save analysis results (S3 bucket or local path)
+save_location: str = str(Path(__file__).parents[3] / "analysis_outputs")
```
Save location is not used in the notebook currently, so okay to drop

mogres replied:
Updated to use save location in [cd1ee6](https://github.com/simularium/subcell-pipeline/pull/58/commits/cd1ee6ac144ac0a9c368e04f2277e544a29e2aad)

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/constants.py:
```diff
@@ -0,0 +1,5 @@
+ABSOLUTE_TOLERANCE: float = 1e-6
```
```suggestion
ABSOLUTE_TOLERANCE: float = 1e-6
"""Tolerance level for ???"""
```

Add docstrings after each of these constants so they show up in the documentation.

mogres replied:
Fixed in [f8ffc9](https://github.com/simularium/subcell-pipeline/pull/58/commits/f8ffc9bc7f8a2e06701ffbdfdb6849aa836ee4b2)

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/constants.py:
```diff
@@ -0,0 +1,5 @@
+SIMULATOR_COLOR_MAP: dict[str, str] = {
+    "readdy": "#ca562c",
+    "cytosim": "#008080",
+}
```
```suggestion
}
"""Map of simulator name to color."""
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/vectors.py:
```diff
@@ -0,0 +1,54 @@
+
+def get_end_to_end_unit_vector(
+    polymer_trace: np.ndarray,
+) -> Tuple[np.array, Union[float, np.floating[Any]]]:
```
If possible, each method should only return one thing. Can you have a separate method handle vector lengths?

mogres replied:
Fixed in [3dd2d6](https://github.com/simularium/subcell-pipeline/pull/58/commits/3dd2d6e1d02df67ab33d08977268c6dfeb5911f3)

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/plot_compression_metrics.py:
```diff
@@ -0,0 +1,88 @@
+        Whether to use real time for the x-axis.
+        Defaults to False.
+    """
+    if figure_path is None:
```
Expected behavior would be that is figure path is `None`, then the figures are not saved

mogres replied:
Fixed in [cd1ee6](https://github.com/simularium/subcell-pipeline/pull/58/commits/cd1ee6ac144ac0a9c368e04f2277e544a29e2aad)

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/compression_analysis.py:
```diff
@@ -208,21 +255,27 @@ def get_contour_length_from_trace(
-    **options: dict
+
+    **options
         Additional options as key-value pairs.
```
This docstring is not getting parsed correctly. I think you need a newline between `Additional options as key-value pairs.` and `bending_constant: float`. Alternatively, I think you can just promote the kwargs to actual arguments, so they are directly used:

```python
def get_bending_energy_from_trace(
    polymer_trace: np.ndarray,
    bending_constant: float = 1.0,
) -> Union[float, np.floating[Any]]:
```

It should still work with the generic `functions[self](polymer_trace, **options)` as long as the arguments have a default value.

mogres replied:
Fixed in [f8ffc9](https://github.com/simularium/subcell-pipeline/pull/58/commits/f8ffc9bc7f8a2e06701ffbdfdb6849aa836ee4b2)

I opted to keep the kwargs to have a uniform way of passing arguments to the compression metric calculations. Promoting to actual arguments may cause errors if unexpected keywords are received as input.

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/compression_analysis.py:
```diff
@@ -208,21 +255,27 @@ def get_contour_length_from_trace(
+        bending energy per monomer of the polymer trace
     """
-    bending_constant = options.get("bending_constant", 1)
+    bending_constant = options.get("bending_constant", 1.0)
```
For all instances of default values, define a constant (e.g. `DEFAULT_BENDING_CONSTANT = 1.0`) instead of using magic numbers.

mogres replied:
Fixed in [f8ffc9](https://github.com/simularium/subcell-pipeline/pull/58/commits/f8ffc9bc7f8a2e06701ffbdfdb6849aa836ee4b2)

jessicasyu reviewed (commented):
Looking good so far, mostly structure and doc-related comments. Notebook ran correctly for Cytosim (both loading and re-calculating metrics).

- All the modules (`compression_analysis.py`, `vectors.py`, `constants.py`, and `plot_compression_metrics.py`) need a module docstring (so they are displayed in the doc summary)
- Consider separating the actual metrics calculation methods (e.g. `get_total_fiber_twist`) into a `compression_metrics.py` module and combining the plotting and the actual calculation loop (`get_compression_metric_data` and `calculate_compression_metrics`) in `compression_analysis.py`
- Consider putting the enum in a separate file 
- Consider renaming `COMPRESSIONMETRIC` to `CompressionMetric` for readability
- Need to address the mypy errors when running `mypy subcell_pipeline/analysis/compression_metrics/`

mogres committed:
Apply suggestions from code review

Co-authored-by: Jessica S. Yu <15913767+jessicasyu@users.noreply.github.com>

mogres committed:
Save plots of metrics vs time to analysis outputs

mogres committed:
Fix TOC links

mogres committed:
update docstring and use default bending constant

mogres committed:
* Rename COMPRESSIONMETRIC -> CompressionMetric
* Refactor compression_analysis into separate functions for metric calculation and plotting + analysis
* Add docstrings for modules
* Fix linting errors

mogres commented:
> * All the modules (`compression_analysis.py`, `vectors.py`, `constants.py`, and `plot_compression_metrics.py`) need a module docstring (so they are displayed in the doc summary)

Done!

> * Consider separating the actual metrics calculation methods (e.g. `get_total_fiber_twist`) into a `compression_metrics.py` module and combining the plotting and the actual calculation loop (`get_compression_metric_data` and `calculate_compression_metrics`) in `compression_analysis.py`
> * Consider putting the enum in a separate file

Refactored into `compression_metric` for the enum, `polymer_trace` for metric calculating functions, and `compression_analysis` for the actual calculation loop and plotting.

> * Consider renaming `COMPRESSIONMETRIC` to `CompressionMetric` for readability

Done

> * Need to address the mypy errors when running `mypy subcell_pipeline/analysis/compression_metrics/`

Fixed most type hinting errors. I wasn't sure how to fix a couple so have left a comment explaining the issue and a `# type:ignore` comment for now. `mypy` passes without errors now.

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/_compare_compression_metrics.py:
```diff
@@ -0,0 +1,141 @@
+"""
+## Select metrics to analyze
+
+Available metrics are defined in the `COMPRESSIONMETRIC` enum.
```
```suggestion
Available metrics are defined in the `CompressionMetric` enum.
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/polymer_trace.py:
```diff
@@ -0,0 +1,475 @@
+    :
+        bending energy per monomer of the polymer trace
+    """
+    bending_constant = options.get("bending_constant", DEFAULT_BENDING_CONSTANT)
```
Add 

```python
assert isinstance(bending_constant, float)
```

here to fix the typing error on 214

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/polymer_trace.py:
```diff
@@ -0,0 +1,475 @@
+    """
+    compression_axis = options.get("compression_axis", 0)
+    signed = options.get("signed", True)
+    tolerance = options.get("tolerance", ABSOLUTE_TOLERANCE)
```
Add 

```python
assert isinstance(signed, bool)
assert isinstance(tolerance, float)
```

here to fix the typing error on 257

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/polymer_trace.py:
```diff
@@ -0,0 +1,475 @@
+# Methods to calculate metrics from polymer trace data.
```
Needs to be a docstring and not a comment in order to show up in docs:

```python
"""
Methods to calculate metrics from polymer trace data.
"""
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/vectors.py:
```diff
@@ -0,0 +1,54 @@
+# Methods for vector operations.
```
Needs to be a docstring and not a comment in order to show up in docs:

```python
"""
Methods for vector operations.
"""
```

jessicasyu commented on subcell_pipeline/analysis/compression_metrics/constants.py:
```diff
@@ -0,0 +1,14 @@
+# Constants used in compression metric analysis.
```
Needs to be a docstring and not a comment in order to show up in docs:

```python
"""
Constants used in compression metric analysis.
"""
```

jessicasyu reviewed (approved):
Looks great! Added some suggestions to fix the typing errors and make sure the module docstrings show up.

mogres committed:
Update docstrings to one line

mogres committed:
Fix typo and update typing hints

mogres committed:
Replace COMPRESSIONMETRIC -> CompressionMetric in analysis files

mogres committed:
Use default compression distance from constants

mogres committed:
Add docstring to top level __init__.py
